File: android/src/main/java/SharePlugin.kt
package com.plugin.share

import android.app.Activity
import android.content.Intent
import app.tauri.annotation.Command
import app.tauri.annotation.InvokeArg
import app.tauri.annotation.TauriPlugin
import app.tauri.plugin.JSObject
import app.tauri.plugin.Plugin
import app.tauri.plugin.Invoke

@InvokeArg
class ShareArgs {
    var url: String? = null
    var title: String? = null
    var text: String? = null
}

@TauriPlugin
class SharePlugin(private val activity: Activity): Plugin(activity) {

    @Command
    fun share(invoke: Invoke) {
        val args = invoke.parseArgs(ShareArgs::class.java)

        // Build the share text: combine text and URL if both present
        val shareText = buildString {
            args.text?.let { append(it) }
            args.url?.let { url ->
                if (isNotEmpty()) append("\n")
                append(url)
            }
        }

        if (shareText.isEmpty()) {
            invoke.reject("Nothing to share: provide at least 'url' or 'text'")
            return
        }

        val intent = Intent(Intent.ACTION_SEND).apply {
            type = "text/plain"
            args.title?.let { putExtra(Intent.EXTRA_SUBJECT, it) }
            putExtra(Intent.EXTRA_TEXT, shareText)
        }

        val chooserTitle = args.title ?: "Share"
        val chooser = Intent.createChooser(intent, chooserTitle)

        activity.startActivity(chooser)
        invoke.resolve(JSObject())
    }

    @Command
    fun canShare(invoke: Invoke) {
        // On Android, we can always share text content
        val ret = JSObject()
        ret.put("value", true)
        invoke.resolve(ret)
    }
}

File: ios/Sources/SharePlugin.swift
import SwiftRs
import Tauri
import UIKit
import WebKit

/// Arguments for the share command.
class ShareArgs: Decodable {
    /// URL to share.
    let url: String?
    /// Title for the shared content (may be ignored by some targets).
    let title: String?
    /// Text content to share.
    let text: String?
}

class SharePlugin: Plugin {
    
    @objc public func share(_ invoke: Invoke) {
        do {
            let args = try invoke.parseArgs(ShareArgs.self)
            
            var items: [Any] = []
            
            // Add text if provided
            if let text = args.text {
                items.append(text)
            }
            
            // Add URL if provided
            if let urlString = args.url, let url = URL(string: urlString) {
                items.append(url)
            }
            
            // Reject if nothing to share
            if items.isEmpty {
                invoke.reject("Nothing to share: provide at least 'url' or 'text'")
                return
            }
            
            // Present share sheet on main thread
            DispatchQueue.main.async {
                let activityVC = UIActivityViewController(
                    activityItems: items,
                    applicationActivities: nil
                )
                
                // Get the root view controller to present from
                guard let windowScene = UIApplication.shared.connectedScenes.first as? UIWindowScene,
                      let rootVC = windowScene.windows.first?.rootViewController else {
                    invoke.reject("Could not find root view controller")
                    return
                }
                
                // Handle iPad popover presentation
                if let popover = activityVC.popoverPresentationController {
                    popover.sourceView = rootVC.view
                    popover.sourceRect = CGRect(
                        x: rootVC.view.bounds.midX,
                        y: rootVC.view.bounds.midY,
                        width: 0,
                        height: 0
                    )
                    popover.permittedArrowDirections = []
                }
                
                rootVC.present(activityVC, animated: true) {
                    invoke.resolve()
                }
            }
        } catch {
            invoke.reject("Failed to parse arguments: \(error.localizedDescription)")
        }
    }
    
    @objc public func canShare(_ invoke: Invoke) {
        // iOS always supports sharing
        invoke.resolve(["value": true])
    }
}

@_cdecl("init_plugin_share")
func initPlugin() -> Plugin {
    return SharePlugin()
}
